File: matiasmasso/Mat.NET/DTO/DTOCnap.cs
using System;
using System.Collections.Generic;
using System.Linq;

namespace DTO
{
    public class DTOCnap : DTOBaseGuid
    {
        public string Id { get; set; }
        public DTOCnap Parent { get; set; }
        public List<DTOCnap> Children { get; set; }
        public DTOLangText NomShort { get; set; }
        public DTOLangText NomLong { get; set; }
        public List<string> Tags { get; set; }
        public List<DTOProduct> Products { get; set; }

        public DTOCnap() : base()
        {
            NomShort = new DTOLangText(base.Guid, DTOLangText.Srcs.CnapShort);
            NomLong = new DTOLangText(base.Guid, DTOLangText.Srcs.CnapLong);
        }

        public DTOCnap(Guid oGuid) : base(oGuid)
        {
            NomShort = new DTOLangText(base.Guid, DTOLangText.Srcs.CnapShort);
            NomLong = new DTOLangText(base.Guid, DTOLangText.Srcs.CnapLong);
        }

        public string ShortFullNom(DTOLang oLang)
        {
            string retval = string.Format("{0} {1}", Id, NomShort.Tradueix(oLang));
            return retval;
        }

        public string FullNom(DTOLang oLang)
        {
            string sNom = NomLong.Tradueix(oLang);
            if (sNom == "")
                sNom = NomShort.Tradueix(oLang);
            string retval = string.Format("{0} {1}", Id, sNom);
            return retval;
        }

        public static string FullNom(DTOCnap cnap, DTOLang lang)
        {
            string retval = "";
            if (cnap != null)
            {
                retval = cnap.FullNom(lang);
            }
            return retval;
        }

        public List<DTOCnap> Parents()
        {
            List<DTOCnap> retval = new List<DTOCnap>();
            DTOCnap item = this;
            while (item.Parent != null)
            {
                retval.Add(item.Parent);
                item = item.Parent;
            }
            return retval;
        }
        public List<DTOCnap> Descendants()
        {
            List<DTOCnap> retval = new List<DTOCnap>();
            retval.Add(this);
            foreach (DTOCnap child in this.Children)
            {
                retval.AddRange(child.Descendants());
            }
            return retval;
        }

        public bool IsSelfOrChildOf(DTOCnap oParent)
        {
            bool retval = oParent.Equals(this);
            if (!retval)
                retval = IsChildOf(oParent);
            return retval;
        }

        public bool IsSelfOrChildOfAny(List<DTOCnap> oParents)
        {
            bool retval = oParents.Any(x => x.Equals(this) || this.IsChildOf(x));
            return retval;
        }

        public bool IsChildOf(DTOCnap oParent)
        {
            bool retval = this.Parents().Any(x => x.Equals(oParent));
            return retval;
        }

        public List<DTOProductCategory> MatchingCategories(List<DTOProductCategory> src)
        {
            List<DTOProductCategory> retval = src.Where(x => this.Descendants().Any(y => x.Cnap.Guid.Equals(y.Guid))).ToList();
            return retval;
        }
    }
}

File: matiasmasso/Mat.NET/DT2/DTOCnap.vb
Public Class DTOCnap
    Inherits DTOBaseGuid

    Property Id As String
    Property Parent As DTOCnap
    Property Children As List(Of DTOCnap)
    Property NomShort As DTOLangText
    Property NomLong As DTOLangText
    Property Tags As List(Of String)
    Property Products As List(Of DTOProduct)

    Public Sub New()
        MyBase.New
        _NomShort = New DTOLangText
        _NomLong = New DTOLangText
    End Sub

    Public Sub New(oGuid As Guid)
        MyBase.New(oGuid)
    End Sub

    Public Function ShortFullNom(oLang As DTOLang) As String
        Dim retval As String = String.Format("{0} {1}", _Id, _NomShort.tradueix(oLang))
        Return retval
    End Function

    Public Function FullNom(oLang As DTOLang) As String
        Dim sNom As String = _NomLong.tradueix(oLang)
        If sNom = "" Then sNom = _NomShort.tradueix(oLang)
        Dim retval As String = String.Format("{0} {1}", _Id, sNom)
        Return retval
    End Function

    Public Function Parents() As List(Of DTOCnap)
        Dim retval As New List(Of DTOCnap)
        Dim item As DTOCnap = Me
        Do While item.Parent IsNot Nothing
            retval.Add(item.Parent)
            item = item.Parent
        Loop
        Return retval
    End Function

    Public Function IsSelfOrChildOf(oParent As DTOCnap) As Boolean
        Dim retval As Boolean = oParent.Equals(Me)
        If Not retval Then
            retval = IsChildOf(oParent)
        End If
        Return retval
    End Function

    Public Function IsSelfOrChildOfAny(oParents As List(Of DTOCnap)) As Boolean
        Dim retval As Boolean = oParents.Any(Function(x) x.Equals(Me) OrElse Me.IsChildOf(x))
        Return retval
    End Function

    Public Function IsChildOf(oParent As DTOCnap) As Boolean
        Dim retval As Boolean = Me.Parents.Any(Function(x) x.Equals(oParent))
        Return retval
    End Function


End Class
